### ISSUE TITLE
Can't Start Layer 1 with a proxy server

### ISSUE THREAD

**Issue by Martin-H1** [flynn/flynn#1953]
Hi,

I've read through the documentation and think I'm following the right set of steps to install Flynn, but I can't seem to get layer 1 started.

* I created an Ubuntu 14.04 VM called and added a proxy to /etc/environment to pull code from external repos.

* The program apt-get and update manager use a different proxy setting and will not obey the environment variables. Create a file called 95proxies in /etc/apt/apt.conf.d/, and include the proxy using the correct syntax.

* Reboot to let this take effect.

* Get the Flynn install script:

curl -fsSL -o install-flynn https://dl.flynn.io/install-flynn

Copy it to install-flynn.sh, chmod +x, and make an edit to allow it to go through the proxy by specifying --keyserver-options http-proxy=myproxy

* Run the install script:
 
$ sudo ./install-flynn.sh
 
That results in the happy sight:
 
===> 16:04:22.212 setting libvirt defaults
===> 16:04:22.216 installing Upstart job
===> 16:04:22.241 installation complete!
 
* Start Flynn layer 0 and query its state with these commands
$ sudo start flynn-host
$ sudo status flynn-host
 
Flynn didn’t start so I dumped the log to find out what was going on:

$ sudo cat /var/log/upstart/flynn-host.log
now=2015-10-02T20:34:27+0000 app=host at=start
now=2015-10-02T20:34:27+0000 size=2773551513 at=zpool_size app=host fn=main
DBUG[10-02|16:34:27] zfs get all -Hp flynn-default            package=zfs
16:34:27.956398 host.go:212: could not initialize default provider: exit status 1: "/sbin/zfs zfs get all -Hp flynn-default" => The ZFS modules are not loaded.
Try running '/sbin/modprobe zfs' as root to load them.
 
* I followed the instructions and it worked:
$ sudo /sbin/modprobe zfs
$ sudo start flynn-host
flynn-host start/running, process 19378
$ sudo status flynn-host
flynn-host start/running, process 19378

* Obtain a cluster token.

$ sudo flynn-host init --init-discovery
https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
 
* Init the other cluster members to the same ID using this command:
sudo flynn-host init --discovery https://discovery.flynn.io/clusters/53e8402e-030f-4861-95ba-d5b5a91b5902

* To spoof the DNS records I plan to use a combination of a hosts file and dnsmasq during an evaluation. xip.io is not reachable behind my corporate firewall, so it is not an option.

$ sudo vi /etc/hosts
127.0.0.1            demo.localflynn.com

sudo apt-get dnsmasq
sudo vi /etc/dnsmasq.conf and inserted this line: address=/demo.localflynn.com/127.0.0.1
sudo reboot

now ping martin.demo.localflynn.com gets a response.

* The next step is to start layer 1 using the bootstrap command. So I did that via this command:
$ sudo su
# CLUSTER_DOMAIN=demo.localflynn.com \
    flynn-host bootstrap \
    --min-hosts 3 \
    --discovery https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
 
But at the preset time that times out waiting for the hosts to come online:
12:57:47.557283 check online-hosts
12:58:18.853217 check online-hosts error: timed out waiting for 3 hosts to come online (currently 0 online)

I'm at a bit of a loss as I think I've done everything right, but I don't know what I'm missing. Any help is appreciated.

Martin

**Comment by lmars** [comment]
You have referenced two different cluster discovery URLs:

```
$ sudo flynn-host init --init-discovery
https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
```

```
$ flynn-host init --discovery https://discovery.flynn.io/clusters/53e8402e-030f-4861-95ba-d5b5a91b5902
```

They need to be the same URL.

**Comment by Martin-H1** [comment]
OK. That's a mistake on the second and third node, and it will cause a problem discovering those nodes. So I reissued the command on those nodes:
$ sudo flynn-host init --discovery  https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000 

But I still get the same problem, and even trying to bring up a single node cluster fails:

$ sudo CLUSTER_DOMAIN=demo.localflynn.com     flynn-host bootstrap     --min-hosts 1     --discovery https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
13:38:34.779280 check online-hosts
13:39:05.773819 check online-hosts error: timed out waiting for 1 hosts to come online (currently 0 online)

Just to make sure it wasn't a firewall issue I made sure localhost had full access:
$ sudo ufw allow from 127.0.0.1

So while that cut and paste error is a problem, I don't think it is the root cause.

**Comment by lmars** [comment]
Ok, please run `sudo flynn-host collect-debug-info` on each node in your cluster to upload some logs and system information to a gist, then post the links here along with the contents of `/etc/flynn/host.json` from each node.

**Comment by Martin-H1** [comment]
Thanks for the reply. Let's focus on a single node cluster since I think once that works I can focus on the other.

Node flynn-1
$ cat /etc/flynn/host.json
{
        "args": [
                "--discovery",
                "https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000"
        ]
}

Output from collect-debug-info: https://gist.github.com/anonymous/225f64ebbc3186db7b4c

**Comment by josephglanville** [comment]
I think this might be an HTTP proxy problem. Are your nodes behind a firewall that is preventing them from talking to the discovery service?

**Comment by Martin-H1** [comment]
Yes, we are behind a corporate firewall. As I said in the steps above I established proxies to allow packages to download and scripts to execute. Does flynn require proxy configuration?

**Comment by josephglanville** [comment]
It seems that `flynn-host` will not source `/etc/environment` if it's launched via upstart.

To test you could run flynn-host as root with the proxy variables correctly setup, if that works you could attempt the following.

You could work around this by modifying `/etc/init/flynn-host.conf` to read like so:
```
description "Flynn layer 0"

#start on (started libvirt-bin and started networking)
stop on stopping libvirt-bin
#respawn
#respawn limit 1000 60

script
  . /etc/environment
  /usr/local/bin/flynn-host daemon
end script
```

**Comment by Martin-H1** [comment]
I tried everything as root and verified that my proxy env vars are available to root.

# start flynn-host
flynn-host start/running, process 3608
# status flynn-host
flynn-host start/running, process 3608
# env | grep -i proxy
NO_PROXY=localhost,127.0.0.1,*.mydomain.com,*.localflynn.com
http_proxy=http://http.proxy.mydomain.com:8000
HTTPS_PROXY=http://http.proxy.mydomain.com:8000
https_proxy=http://http.proxy.mydomain.com:8000
no_proxy=localhost,127.0.0.1,*.mydomain.com,*.localflynn.com
HTTP_PROXY=http://http.proxy.mydomain.com:8000
# CLUSTER_DOMAIN=demo.localflynn.com     flynn-host bootstrap     --min-hosts 1     --discovery https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
14:27:16.772749 check online-hosts
14:27:47.547826 check online-hosts error: timed out waiting for 1 hosts to come online (currently 0 online)

Same issue. Note that mydomain is the masked value for my actual domain. During setup I used curl and apt-get, so I believe these variables are correct.

**Comment by Martin-H1** [comment]
The funky fonts above are caused by the pound sign which is the root command prompt triggering markdown.

**Comment by josephglanville** [comment]
@Martin-H1 can you re-run `sudo flynn-host collect-debug-info`

**Comment by Martin-H1** [comment]
Done. Note I verified flynn-host is running, but I noticed a peculiar connection refused to localhost port 1111 and then it is not running:
```text
# status flynn-host
flynn-host start/running, process 4794
root@vc2crtpb064509n:/home/a231861# flynn-host collect-debug-info
INFO[10-07|14:42:17] uploading logs and debug information to a private, anonymous gist
INFO[10-07|14:42:17] this may take a while depending on the size of your logs
INFO[10-07|14:42:17] getting flynn-host logs
INFO[10-07|14:42:17] getting job logs
EROR[10-07|14:42:47] error getting job logs, falling back to on-disk logs err="could not list hosts: Get http://127.0.0.1:1111/services/flynn-host/instances: dial tcp 127.0.0.1:1111: connection refused"
INFO[10-07|14:42:47] getting system information
EROR[10-07|14:42:47] error capturing output of "bash -c tail -n 1000 /var/log/flynn/**/*.log" err="exit status 1"
INFO[10-07|14:42:47] creating anonymous gist
32.51 KB / 32.51 KB [====================================] 100.00 % 7.35 KB/s 4s
INFO[10-07|14:42:51] debug information uploaded to: https://gist.github.com/dcff2c2935e894792f72
root@vc2crtpb064509n:/home/a231861# status flynn-host
flynn-host stop/waiting
```

Odd, I don't have any firewall rules that should block it:
```text
# ufw status
Status: active

To                         Action      From
--                         ------      ----
2783/tcp                   ALLOW       Anywhere
2703/tcp                   ALLOW       Anywhere
22                         ALLOW       Anywhere
Anywhere                   ALLOW       127.0.0.1
```

**Comment by Martin-H1** [comment]
Out of curiosity I made the changes above to /etc/init/flynn-host.conf and tried again. It made no difference in the behavior of Flynn.

**Comment by josephglanville** [comment]
@Martin-H1 it would be good if you could post the log of that attempt too. `collect-debug-info` will do the trick.

It does seem however like there is something not respecting HTTP proxy variables, this is fairly likely as it's not a common configuration. I will see if I can setup a HTTP proxy to test with.

**Comment by titanous** [comment]
I don't think the Go HTTP client is set up for HTTP proxies.

**Comment by lmars** [comment]
@Martin-H1 given `flynn-host init --init-discovery` works when you run from the command line indicates that your proxy is functioning when you have the environment variables set (i.e. when using a login shell), so the presence of i/o timeouts in your `flynn-host` log probably mean the proxy variables aren't being set for the `flynn-host` daemon.

Upstart has [an env stanza](http://upstart.ubuntu.com/cookbook/#env) so I would try using that to set the `HTTP_PROXY` variable in `/etc/init/flynn-host.conf`:

```
description "Flynn layer 0"

#start on (started libvirt-bin and started networking)
stop on stopping libvirt-bin
#respawn
#respawn limit 1000 60

env HTTP_PROXY=http://mydomain.com

exec /usr/local/bin/flynn-host daemon
```

If that doesn't work, I would try using `/usr/bin/env` to set it:

```
description "Flynn layer 0"

#start on (started libvirt-bin and started networking)
stop on stopping libvirt-bin
#respawn
#respawn limit 1000 60

exec /usr/bin/env HTTP_PROXY=http://mydomain.com /usr/local/bin/flynn-host daemon
```

If that still doesn't work, try starting `flynn-host` from a login shell and bootstrap from another terminal.

**Comment by Martin-H1** [comment]
I got pulled away from this today, but I will try these things tomorrow.

A firewall with an outbound proxy isn't that unusual a configuration and is supported by both Tomcat and Node.js. I agree that if I was able to get the token via the command line it sounds like the proxy is getting lost when running as a service.

**Comment by lmars** [comment]
@Martin-H1 @titanous FYI the default Go HTTP client [supports the HTTP_PROXY environment variable](https://github.com/golang/go/blob/release-branch.go1.4/src/net/http/transport.go#L30-L32).

**Comment by Martin-H1** [comment]
I got a chance to try the first suggestion of adding the environment variables to /etc/init/flynn-host.conf, stopping and starting flynn-host to ensure the changes are consumed, and then trying to start layer 1. It didn't work, but here's the output and link to the gist output:

https://gist.github.com/5769678f0ceff6775b38
 
```text
# cat /etc/init/flynn-host.conf
description "Flynn layer 0"

#start on (started libvirt-bin and started networking)
stop on stopping libvirt-bin
#respawn
#respawn limit 1000 60

env http_proxy=http://http.proxy.mydomain.com:8000
env https_proxy=http://http.proxy.mydomain.com:8000
env no_proxy="localhost,127.0.0.1,*.mydomain.com,*.localflynn.com"
env HTTP_PROXY=http://http.proxy.mydomain.com:8000
env HTTPS_PROXY=http://http.proxy.mydomain.com:8000
env NO_PROXY="localhost,127.0.0.1,*.mydomain.com,*.localflynn.com"

exec /usr/local/bin/flynn-host daemon
# stop flynn-host
flynn-host stop/waiting
root@vc2crtpb064509n:/etc/init# start flynn-host
flynn-host start/running, process 29850
root@vc2crtpb064509n:/etc/init# CLUSTER_DOMAIN=demo.localflynn.com \
>     flynn-host bootstrap \
>     --min-hosts 1 \
>     --discovery https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
16:58:21.542095 check online-hosts

16:59:36.622542 check online-hosts error: timed out waiting for 1 hosts to come online (currently 0 online)
root@vc2crtpb064509n:/etc/init#
root@vc2crtpb064509n:/etc/init# flynn-host collect-debug-info
INFO[10-08|17:02:04] uploading logs and debug information to a private, anonymous gist
INFO[10-08|17:02:04] this may take a while depending on the size of your logs
INFO[10-08|17:02:04] getting flynn-host logs
INFO[10-08|17:02:04] getting job logs
EROR[10-08|17:02:34] error getting job logs, falling back to on-disk logs err="could not list hosts: Get http://127.0.0.1:1111/services/flynn-host/instances: dial tcp 127.0.0.1:1111: connection refused"
INFO[10-08|17:02:34] getting system information
EROR[10-08|17:02:34] error capturing output of "bash -c tail -n 1000 /var/log/flynn/**/*.log" err="exit status 1"
INFO[10-08|17:02:34] creating anonymous gist
36.43 KB / 36.43 KB [======================================] 100.00 % 44.22 KB/s 0
INFO[10-08|17:02:35] debug information uploaded to: https://gist.github.com/5769678f0ceff6775b38
```

**Comment by Martin-H1** [comment]
I contacted a team member who's more of a Linux guru than me. He said that my editing /etc/environment and rebooting should have been adequate to provide a serivce with the environment variable. This confirmed by my ability to use curl from a root shell to install Flynn initially.

I tried the other suggestion of adding /usr/bin/env HTTP_PROXY to the exec in /etc/init/flynn-host.conf but it didn't help either. The results are below:

```text
# cat /etc/init/flynn-host.conf
description "Flynn layer 0"

#start on (started libvirt-bin and started networking)
stop on stopping libvirt-bin
#respawn
#respawn limit 1000 60

env http_proxy=http://http.proxy.mydomain.com:8000
env https_proxy=http://http.proxy.mydomain.com:8000
env no_proxy="localhost,127.0.0.1,*.mydomain.com,*.localflynn.com"
env HTTP_PROXY=http://http.proxy.mydomain.com:8000
env HTTPS_PROXY=http://http.proxy.mydomain.com:8000
env NO_PROXY="localhost,127.0.0.1,*.mydomain.com,*.localflynn.com"

exec /usr/bin/env HTTP_PROXY=http://http.proxy.mydomain.com:8000 /usr/local/bin/flynn-host daemon
# stop flynn-host
flynn-host stop/waiting
# start flynn-host
flynn-host start/running, process 24518
# status flynn-host
flynn-host start/running, process 24518
# CLUSTER_DOMAIN=demo.localflynn.com \
>     flynn-host bootstrap \
>     --min-hosts 1 \
>     --discovery https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
10:15:32.607764 check online-hosts
10:16:47.737606 check online-hosts error: timed out waiting for 1 hosts to come online (currently 0 online)
#
# flynn-host collect-debug-info
INFO[10-09|10:19:56] uploading logs and debug information to a private, anonymous gist
INFO[10-09|10:19:56] this may take a while depending on the size of your logs
INFO[10-09|10:19:56] getting flynn-host logs
INFO[10-09|10:19:56] getting job logs
EROR[10-09|10:20:25] error getting job logs, falling back to on-disk logs err="could not list hosts: Get http://127.0.0.1:1111/services/flynn-host/instances: dial tcp 127.0.0.1:1111: connection refused"
INFO[10-09|10:20:25] getting system information
EROR[10-09|10:20:25] error capturing output of "bash -c tail -n 1000 /var/log/flynn/**/*.log" err="exit status 1"
INFO[10-09|10:20:25] creating anonymous gist
35.21 KB / 35.21 KB [======================================] 100.00 % 66.17 KB/s 0
INFO[10-09|10:20:26] debug information uploaded to: https://gist.github.com/b6a964f72c541a0c793e
```

**Comment by titanous** [comment]
There are no errors in the `flynn-host daemon` log from your last run, and the registration showed up successfully. I wonder if `flynn-host bootstrap` is failing to connect to `http://10.96.43.98:1113` due to the proxy settings?

**Comment by Martin-H1** [comment]
It should get those values from the /etc/environment edit which allows curl to work. I verified this with the following:

```text
# env | grep -i proxy
NO_PROXY=localhost,127.0.0.1,*.mydomain.com,*.localflynn.com
http_proxy=http://http.proxy.mydomain.com:8000
HTTPS_PROXY=http://http.proxy.mydomain.com:8000
https_proxy=http://http.proxy.mydomain.com:8000
no_proxy=localhost,127.0.0.1,*.mydomain.com,*.localflynn.com
HTTP_PROXY=http://http.proxy.mydomain.com:8000
root@vc2crtpb064509n:/home/a231861# CLUSTER_DOMAIN=demo.localflynn.com     flynn-host bootstrap     --min-hosts 1     --discovery https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
10:35:20.120949 check online-hosts
10:36:35.407480 check online-hosts error: timed out waiting for 1 hosts to come online (currently 0 online)
```

**Comment by titanous** [comment]
Yeah, what I'm wondering is if round-tripping to `10.96.43.98` through the proxy will work? It's not in `NO_PROXY` that I can see.

**Comment by Martin-H1** [comment]
I can add that to no_proxy, but that's the IP address of localhost. It would be pretty weird to use a proxy to reach yourself.

**Comment by titanous** [comment]
I don't think the Go client is doing an inventory of network interfaces and trying to work out which one maps to localhost and whitelisting those IPs. Is that standard behavior for other HTTP clients that use proxies? (eg. what happens when you `curl http://10.96.43.98:1113`?)

**Comment by Martin-H1** [comment]
I think you are on to something:
```text
# curl http://10.96.43.98:1113?
curl: (56) Recv failure: Connection reset by peer
```

**Comment by Martin-H1** [comment]
It got farther with the modified no_proxy, but then got this error:
wait postgres-wait error: bootstrap: timed out waiting for http://postgres-api.discoverd/ping, last response 404
```text
# env | grep -i no_proxy                        NO_PROXY=localhost,127.0.0.1,10.96.43.98,*.fmr.com,*.localflynn.com
no_proxy=localhost,127.0.0.1,10.96.43.98,*.fmr.com,*.localflynn.com
root@vc2crtpb064509n:/home/a231861# CLUSTER_DOMAIN=demo.localflynn.com     flynn-host bootstrap     --min-hosts 1     --discovery https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
10:53:15.760131 check online-hosts
10:53:16.217380 require-env require-env
10:53:16.233992 run-app flannel
10:53:18.443564 run-app discoverd
10:53:19.507556 wait-hosts wait-hosts
10:53:26.517536 gen-random pg-password
10:53:26.517686 gen-random pg-password a52fe11f0e507b05a1879fe0c44e0c36
10:53:26.517692 run-app postgres
10:53:28.522047 gen-random controller-key
10:53:28.522183 gen-random controller-key 7aca91ae5491e64cffad765768004097
10:53:28.522190 gen-random dashboard-session-secret
10:53:28.522231 gen-random dashboard-session-secret e70080fd9ea149112373db441fd7c705
10:53:28.522236 gen-random dashboard-login-token
10:53:28.522271 gen-random dashboard-login-token e6594a3d9a8beb98a6fc45528e6a6f9b
10:53:28.522276 gen-random name-seed
10:53:28.522301 gen-random name-seed 927284ad4676275859d0
10:53:28.522305 gen-random router-sticky-key
10:53:28.522349 gen-random router-sticky-key fwEX/u6xePsFFO1Z+DXVDA5UmRu5Vkloz+M54bDsHZQ=
10:53:28.522353 wait postgres-wait
10:55:28.601403 wait postgres-wait error: bootstrap: timed out waiting for http://postgres-api.discoverd/ping, last response 404
```

**Comment by titanous** [comment]
Okay, that failure is likely caused by trying to communicate with the container network through the proxy. Add `*.discoverd` and `100.100.*` to the `NO_PROXY` list.

Bootstrap isn't currently idempotent, so do the following to restart before trying bootstrap again:

```
stop flynn-host
rm /var/lib/flynn/host-state.bolt
start flynn-host
```

**Comment by Martin-H1** [comment]
Added that but it looks like the same symptom.
```text
# export NO_PROXY=localhost,127.0.0.1,10.96.43.98,*.mydomain.com,*.localflynn.com,*.discoverd,100.100.*
root@vc2crtpb064509n:/home/a231861# export no_proxy=localhost,127.0.0.1,10.96.43.98,*.mydomain.com,*.localflynn.com,*.discoverd,100.100.*
root@vc2crtpb064509n:/home/a231861# stop flynn-host
stop: Unknown instance:
root@vc2crtpb064509n:/home/a231861# rm /var/lib/flynn/host-state.bolt
root@vc2crtpb064509n:/home/a231861# start flynn-host
flynn-host start/running, process 28014
root@vc2crtpb064509n:/home/a231861# CLUSTER_DOMAIN=demo.localflynn.com     flynn-host bootstrap     --min-hosts 1     --discovery https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
11:08:43.525433 check online-hosts
11:08:43.665466 require-env require-env
11:08:43.665744 run-app flannel
11:08:44.195365 run-app discoverd
11:08:44.793834 wait-hosts wait-hosts
11:08:47.302553 gen-random pg-password
11:08:47.302745 gen-random pg-password d8a6404ebc6506e455dd664b334b2177
11:08:47.302751 run-app postgres
11:08:48.830044 gen-random controller-key
11:08:48.830219 gen-random controller-key 54bac496013637accf09fc4524977cf8
11:08:48.830226 gen-random dashboard-session-secret
11:08:48.830271 gen-random dashboard-session-secret 8492fd19fc6e7349ea0770ed4eb6691e
11:08:48.830276 gen-random dashboard-login-token
11:08:48.830311 gen-random dashboard-login-token 1dbd397bbf26d7cb3e3a3582f1feb02c
11:08:48.830315 gen-random name-seed
11:08:48.830350 gen-random name-seed dbd88ff0e06b241139b2
11:08:48.830354 gen-random router-sticky-key
11:08:48.830403 gen-random router-sticky-key vnQjJtaGHteooYKcwaiNrE0NwEGrGg7sRssun31Ks24=
11:08:48.830407 wait postgres-wait
11:10:48.864619 wait postgres-wait error: bootstrap: timed out waiting for http://postgres-api.discoverd/ping, last response 404
```

**Comment by titanous** [comment]
Okay, please post new debug info.

**Comment by Martin-H1** [comment]
Done, see https://gist.github.com/f91138f2072e6702c5d3

**Comment by titanous** [comment]
The error from the postgres appliance is:

```text
Get http://10.96.43.98:1111/services/postgres: dial tcp 10.96.43.98:1111: i/o timeout
```

It looks like there is a ufw firewall configured that might be blocking communication between containers. We can sort out how to fix that, but for testing purposes can you disable ufw?

**Comment by Martin-H1** [comment]
Yes, I can try that.

**Comment by Martin-H1** [comment]
Interesting, the error changed from a 404 to 503. debug information uploaded to: https://gist.github.com/236f0977a77d7e87001f

```text
# ufw disable
Firewall stopped and disabled on system startup
# stop flynn-host
stop: Unknown instance:
# rm /var/lib/flynn/host-state.bolt
# start flynn-host
flynn-host start/running, process 29652
# CLUSTER_DOMAIN=demo.localflynn.com \
>     flynn-host bootstrap \
>     --min-hosts 1 \
>     --discovery https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
11:30:49.345988 check online-hosts
11:30:49.459577 require-env require-env
11:30:49.459840 run-app flannel
11:30:49.995625 run-app discoverd
11:30:50.663520 wait-hosts wait-hosts
11:30:52.166786 gen-random pg-password
11:30:52.166959 gen-random pg-password 57ba4cb9727e2fca109a6f48788eceb4
11:30:52.166967 run-app postgres
11:30:53.729975 gen-random controller-key
11:30:53.730145 gen-random controller-key fcc8e3b387ab67a0bb33a259dcb8901e
11:30:53.730152 gen-random dashboard-session-secret
11:30:53.730183 gen-random dashboard-session-secret 75410b3a3f237b2797fd4d4a08a00010
11:30:53.730195 gen-random dashboard-login-token
11:30:53.730231 gen-random dashboard-login-token 9113111fb62fb6611abe2f4f18496f34
11:30:53.730236 gen-random name-seed
11:30:53.730261 gen-random name-seed 52514dfabc42af257f66
11:30:53.730266 gen-random router-sticky-key
11:30:53.730338 gen-random router-sticky-key h3vBq70JlelJXXqlkdBIsWP/fNiTcIbfg8W86zCKPlk=
11:30:53.730356 wait postgres-wait
11:32:18.494871 wait postgres-wait error: bootstrap: timed out waiting for http://postgres-api.discoverd/ping, last response 503
# flynn-host collect-debug-info
INFO[10-09|11:32:59] uploading logs and debug information to a private, anonymous gist
INFO[10-09|11:32:59] this may take a while depending on the size of your logs
INFO[10-09|11:32:59] getting flynn-host logs
INFO[10-09|11:32:59] getting job logs
INFO[10-09|11:32:59] getting system information
INFO[10-09|11:32:59] creating anonymous gist
172.03 KB / 172.03 KB [===================================] 100.00 % 251.93 KB/s 0
INFO[10-09|11:33:00] debug information uploaded to: https://gist.github.com/236f0977a77d7e87001f
```

**Comment by titanous** [comment]
Indeed, we've gotten quite a bit further. As far as I can see the logs on the daemon/container side look good and postgres appears to have started successfully. What happens when you run `curl http://100.100.17.2/ping`?

**Comment by Martin-H1** [comment]
```text
# curl http://100.100.17.2/ping
<HTML><HEAD>
<TITLE>HTTP 503: Service Unavailable </TITLE>
</HEAD>
<BODY>
<FONT face="Helvetica">
<big><strong></strong></big><BR>
</FONT>
<blockquote>
<TABLE border=0 cellPadding=1 width="80%">
<TR><TD>
<FONT face="Helvetica">
<big>HTTP 503: Service Unavailable  (tcp_error)</big>
<BR>
<BR>
</FONT>
</TD></TR>
<TR><TD>
<FONT face="Helvetica">
HTTP 503: A communication error occurred: "Operation timed out".
</FONT>
</TD></TR>
<TR><TD>
<FONT face="Helvetica">
The Web Server may be down, too busy, or experiencing other problems preventing it from responding to requests. You may wish to try again at a later time.
</FONT>
</TD></TR>
<TR><TD>
<FONT face="Helvetica" SIZE=2>
<BR>
For assistance, please contact tech support @ ###-###-####.
</FONT>
</TD></TR>
</TABLE>
</blockquote>
</FONT>
</BODY></HTML>
```

**Comment by titanous** [comment]
That page is not from Flynn. Do you use the `100.100.0.0/16` IP block on your network?

**Comment by Martin-H1** [comment]
Not sure. I'll ask around.

**Comment by Martin-H1** [comment]
That page is coming from the HTTP proxy. In reading about the no_proxy setting here:

http://www.gnu.org/software/wget/manual/html_node/Proxies.html
http://unix.stackexchange.com/questions/23452/set-a-network-range-in-the-no-proxy-environment-variable 

It supports leading wildcards for domains, but not trailing wildcards on IP addresses. So *.mydomain.com doesn't proxy for any host in my domain, but 100.100.* doesn't prevent going to the proxy for host 100.100.17.2. You need to explicitly enumerate the addresses. So I added that address to the no_proxy setting and reissued the curl command:

```text
# curl http://100.100.17.2/ping
curl: (7) Failed to connect to 100.100.17.2 port 80: No route to host
```

**Comment by josephglanville** [comment]
That is strange. Did you re-bootstrap since the last log you posted?

You should have a local route to this address via `flannel.1`.

**Comment by Martin-H1** [comment]
I stopped flynn-host, removed that file, and restarted before trying. Here's the full list of steps:

```text
# export NO_PROXY=localhost,127.0.0.1,10.96.43.98,*.mydomain.com,*.localflynn.com,*.discoverd,100.100.17.2,100.100.*
# export no_proxy=localhost,127.0.0.1,10.96.43.98,*.mydomain.com,*.localflynn.com,*.discoverd,100.100.17.2,100.100.*
# stop flynn-host
flynn-host stop/waiting
# rm /var/lib/flynn/host-state.bolt
# start flynn-host
flynn-host start/running, process 1013
# CLUSTER_DOMAIN=demo.localflynn.com \
>     flynn-host bootstrap \
>     --min-hosts 1 \
>     --discovery https://discovery.flynn.io/clusters/9bae27e6-1cb2-4486-8791-d6d9dcf00000
12:19:31.759735 check online-hosts
12:19:31.868698 require-env require-env
12:19:31.869168 run-app flannel
12:19:32.434410 run-app discoverd
12:19:33.122972 wait-hosts wait-hosts
12:19:35.627712 gen-random pg-password
12:19:35.627840 gen-random pg-password aa791e6f8ed626a893fa9ea53a5ce897
12:19:35.627846 run-app postgres
12:19:37.495347 gen-random controller-key
12:19:37.495511 gen-random controller-key 452eaaa2cb15b14561a77eae0408401b
12:19:37.495518 gen-random dashboard-session-secret
12:19:37.495543 gen-random dashboard-session-secret 38586dcecb0280a46787e25e5a894dfe
12:19:37.495548 gen-random dashboard-login-token
12:19:37.495594 gen-random dashboard-login-token 1f29435f9b62a8003fb6a770f243f019
12:19:37.495599 gen-random name-seed
12:19:37.495628 gen-random name-seed 9d729c0dea576c5a32aa
12:19:37.495632 gen-random router-sticky-key
12:19:37.495675 gen-random router-sticky-key blF3GTiPEsFY9A6CArtFDM/l9jAv3qmh5n9kWDoTkwM=
12:19:37.495680 wait postgres-wait
12:20:54.011336 wait postgres-wait error: bootstrap: timed out waiting for http://postgres-api.discoverd/ping, last response 503
# curl http://100.100.17.2/ping
curl: (7) Failed to connect to 100.100.17.2 port 80: No route to host
# env | grep -i proxy
NO_PROXY=localhost,127.0.0.1,10.96.43.98,*.mydomain.com,*.localflynn.com,*.discoverd,100.100.17.2,100.100.*
http_proxy=http://http.proxy.mydomain.com:8000
HTTPS_PROXY=http://http.proxy.mydomain.com:8000
https_proxy=http://http.proxy.mydomain.com:8000
no_proxy=localhost,127.0.0.1,10.96.43.98,*.mydomain.com,*.localflynn.com,*.discoverd,100.100.17.2,100.100.*
HTTP_PROXY=http://http.proxy.mydomain.com:8000
```

**Comment by josephglanville** [comment]
Seems to be something very strange going on here.

Mainly
`12:20:54.011336 wait postgres-wait error: bootstrap: timed out waiting for http://postgres-api.discoverd/ping, last response 503`
Implies bootstrap was able to reach `100.100.17.2` but received a 503, yet your shell gets the EHOSTUNREACH meaning that the IP must be down or container is not ARPing properly or something.

**Comment by titanous** [comment]
The 503 is from the proxy, the Postgres API will never return a 503.

**Comment by josephglanville** [comment]
Aye, meaning `flynn-host` daemon didn't actually get the NO_PROXY settings.

Can you post current contents of `/etc/init/flynn-host.conf` and make sure it's env settings mirror what you are using in your shell?

**Comment by Martin-H1** [comment]
D'oh I updated the shell to include 100.100.17.2, but didn't update flynn-host.conf. I'll make that edit and try again.

**Comment by titanous** [comment]
The IP subnet used by flannel will be different each time.

**Comment by Martin-H1** [comment]
That sounds like a problem. It seems like Flynn needs the http_proxy to register a node for discovery, and again to find the list of nodes eligible for a cluster. Once a node has a list of IP addresses it uses HTTP to communicate between them by IP address. However, the http_proxy setting kicks it and no_proxy has only limited support for IP addresses. Since the proxy doesn’t know about our internal IP addresses or the 100.100 subnet, the only way to get this to work is to put all possible IP addressees in no_proxy which seems unworkable. Conversely not setting an http_proxy is unworkable as it prevents the initial discovery of cluster nodes.

Do I understand this properly?

**Comment by josephglanville** [comment]
I think you might be able to avoid the cluster discovery by using `--peer-ips` for bootstrap as shown in help:
```
--peer-ips=IPLIST    use IP address list to connect to cluster
```

You may need to remove the discovery information in the config file for that to work but it could work around the need for HTTP proxy for `flynn-host`.

**Comment by Martin-H1** [comment]
Is it possible to locally host the discovery agent behind a firewall and tell Flynn to use it instead of the global one?

**Comment by lmars** [comment]
@Martin-H1 here is the discovery app: https://github.com/flynn/flynn-discovery

It requires a PostgreSQL database with the following schema: https://github.com/flynn/flynn-discovery/blob/master/schema.sql

If you boot a single node Flynn cluster inside your firewall, then you can run the app there by just pushing it (`flynn create discovery && git push flynn master`), [provisioning a PostgreSQL database](https://flynn.io/docs/postgres) and creating the schema.

When running `flynn-host init` and `flynn-host bootstrap`, set `DISCOVERY_SERVER` to the base URI of your deployed discovery service (e.g. `https://discovery.my-cluster.com`).

**Comment by Martin-H1** [comment]
OK Thanks. It sounds like booting a single node cluster without discovery is a prerequisite anyway, so I'll focus on that.

**Comment by titanous** [comment]
Discovery is not required for a single node cluster (drop all discovery flags).  The problem is whitelisting the container subnets so that the proxy is not used by bootstrap for local communication.

**Comment by Martin-H1** [comment]
I removed the proxy information and tried a single node cluster without discovery using the --peer-ips switch. Here's the output:

```text
# env | grep -i proxy
# status flynn-host
flynn-host stop/waiting
# start flynn-host
flynn-host start/running, process 4992
# status flynn-host
flynn-host start/running, process 4992
# CLUSTER_DOMAIN=demo.localflynn.com \
>     flynn-host bootstrap \
>     --min-hosts 1 \
>     --peer-ips=10.96.43.98
14:49:52.441315 check online-hosts
14:50:22.457376 check online-hosts error: timed out waiting for 1 hosts to come online (currently 0 online)
```
At this point I tried a debug upload but it was blocked so I had to add the proxy back:
```text
# flynn-host collect-debug-info               INFO[10-09|14:56:27] uploading logs and debug information to a private, anonymous gist
INFO[10-09|14:56:27] this may take a while depending on the size of your logs
INFO[10-09|14:56:27] getting flynn-host logs
INFO[10-09|14:56:27] getting job logs
EROR[10-09|14:56:57] error getting job logs, falling back to on-disk logs err="could not list hosts: Get http://127.0.0.1:1111/services/flynn-host/instances: dial tcp 127.0.0.1:1111: connection refused"
INFO[10-09|14:56:57] getting system information
INFO[10-09|14:56:57] creating anonymous gist
443.39 KB / 443.39 KB [=================================] 100.00 % 515.95 KB/s 0
INFO[10-09|14:56:58] debug information uploaded to: https://gist.github.com/0e2d2e04edded66bcaad
```

However, I realized that perhaps the contents of /etc/flynn/host.json might be causing problems, so I removed that file, cleared out the proxy info, and tried again:

```text
1# env | grep -i proxy
root@vc2crtpb064509n:/home/a231861# CLUSTER_DOMAIN=demo.localflynn.com \
>     flynn-host bootstrap \
>     --min-hosts 1 \
>     --peer-ips=10.96.43.98
15:10:45.095950 check online-hosts
15:10:45.097124 require-env require-env
15:10:45.097594 run-app flannel
15:10:45.936392 run-app discoverd

15:10:46.965789 wait-hosts wait-hosts
15:10:46.966880 gen-random pg-password
15:10:46.967010 gen-random pg-password 17496545bb51eaeca3e608952d546041
15:10:46.967016 run-app postgres
15:10:49.424397 gen-random controller-key
15:10:49.424525 gen-random controller-key d5ad0bcbfd62d0b0bd96a27b33f07a62
15:10:49.424532 gen-random dashboard-session-secret
15:10:49.424556 gen-random dashboard-session-secret 385b37ae11d9589aca904fb94b3894a2
15:10:49.424560 gen-random dashboard-login-token
15:10:49.424583 gen-random dashboard-login-token 4203b24bf897bd04445828bedb82f301
15:10:49.424588 gen-random name-seed
15:10:49.424661 gen-random name-seed da0d21115bf7deafb1c6
15:10:49.424666 gen-random router-sticky-key
15:10:49.424713 gen-random router-sticky-key p6bJIY25OAr12IzahynJMjn1vHFGkPEV2FNHFOE7cgM=
15:10:49.424719 wait postgres-wait
15:12:49.511060 wait postgres-wait error: bootstrap: timed out waiting for http://postgres-api.discoverd/ping, last response "Get http://postgres-api.discoverd/ping: dial tcp: lookup postgres-api.discoverd: no such host"
```

That looks better, but the last GET looks like it will be blocked without the configuration of an http_proxy. In any event I added the proxy info back again and did the debug upload.

```text
# flynn-host collect-debug-info
INFO[10-09|15:16:15] uploading logs and debug information to a private, anonymous gist
INFO[10-09|15:16:15] this may take a while depending on the size of your logs
INFO[10-09|15:16:15] getting flynn-host logs
INFO[10-09|15:16:15] getting job logs
INFO[10-09|15:16:15] getting system information
INFO[10-09|15:16:15] creating anonymous gist
623.19 KB / 623.19 KB [=================================] 100.00 % 737.59 KB/s 0
INFO[10-09|15:16:16] debug information uploaded to: https://gist.github.com/7ab62c26a24816a3b59a
```

**Comment by titanous** [comment]
Did you remove `/var/lib/flynn/host-state.bolt` before starting `flynn-host` in you last try? It looks like it was not removed.

**Comment by Martin-H1** [comment]
OK I was away and couldn't work on this, but removing the host-state.bolt seems to have allowed a single node cluster to start:

```text
$ sudo status flynn-host
flynn-host start/running, process 6268
$ sudo stop flynn-host
flynn-host stop/waiting
$ sudo rm /var/lib/flynn/host-state.bolt
$ sudo start flynn-host
flynn-host start/running, process 29888
$ sudo status flynn-host
flynn-host start/running, process 29888
$ sudo CLUSTER_DOMAIN=demo.localflynn.com \
>      flynn-host bootstrap \
>      --min-hosts 1 \
>      --peer-ips=10.96.43.98
15:40:27.169779 check online-hosts
15:40:27.171614 require-env require-env
15:40:27.172129 run-app flannel
15:40:28.381307 run-app discoverd
15:40:29.453979 wait-hosts wait-hosts
15:40:31.457676 gen-random pg-password
15:40:31.457825 gen-random pg-password 8302ac95842b2677382f023be72b78f9
15:40:31.457832 run-app postgres
15:40:33.800397 gen-random controller-key
15:40:33.800550 gen-random controller-key 7b2c7bd3b3302d580947c9387364d718
15:40:33.800558 gen-random dashboard-session-secret
15:40:33.800605 gen-random dashboard-session-secret 2ea7a86374e7fd1dff7f9f5cc3af6619
15:40:33.800611 gen-random dashboard-login-token
15:40:33.800637 gen-random dashboard-login-token a983d1387bad29911b9f8da22e1fa77e
15:40:33.800642 gen-random name-seed
15:40:33.800680 gen-random name-seed e13810d11140e299d55a
15:40:33.800684 gen-random router-sticky-key
15:40:33.800755 gen-random router-sticky-key 1LIYK7jzdksmgUZQpAN+8aMJL0Q2f955PkIg3jvJJlQ=
15:40:33.800761 wait postgres-wait
15:40:41.951322 gen-tls-cert controller-cert
15:40:42.933757 gen-tls-cert controller-cert pin: wOixCkDrIxotWkcvy0FD+w/PI1aWa1qbpHCZtUn0n9s=
15:40:42.933799 run-app controller
15:40:45.790632 wait controller-wait
15:40:48.200768 add-app controller-inception
15:40:48.312167 add-app postgres-app
15:40:48.358295 add-app flannel-app
15:40:48.399899 add-app discoverd-app
15:40:48.455215 scale-app scheduler-scale
15:40:48.465719 run-app scheduler
15:40:49.280606 deploy-app blobstore
15:40:54.780090 deploy-app router
15:40:58.710288 deploy-app gitreceive
15:40:58.805158 wait router-wait
15:41:02.320561 add-route gitreceive-route
15:41:02.342145 add-route controller-route
15:41:02.389480 wait controller-route-wait
15:41:02.910851 deploy-app logaggregator
15:41:03.004593 deploy-app taffy
15:41:03.152332 deploy-app dashboard
15:41:03.346217 add-route dashboard-route
15:41:03.371561 deploy-app status
15:41:03.537588 add-route status-route
15:41:03.587754 wait blobstore-wait
15:41:03.591787 wait gitreceive-wait
15:41:03.592225 wait status-wait
15:41:07.395645 log log-complete
15:41:07.395814 log log-complete

Flynn bootstrapping complete. Install the Flynn CLI (see https://flynn.io/docs/cli for instructions) and paste the line below into a terminal window:

flynn cluster add -p wOixCkDrIxotWkcvy0FD+w/PI1aWa1qbpHCZtUn0n9s= default demo.localflynn.com 7b2c7bd3b3302d580947c9387364d718

The built-in dashboard can be accessed at http://dashboard.demo.localflynn.com and your login token is a983d1387bad29911b9f8da22e1fa77e
```

Now I should be able to create a multi-node by explicitly stating all IP addresses on the command line?

**Comment by titanous** [comment]
Yeah, it looks like specifying the IPs should work as long as the proxy variables are not set and discovery is not enabled.

**Comment by Martin-H1** [comment]
I'm not sure if I should close this issue and start a new one, but the lack of a proxy seems to come back to haunt me when I try to push an app into flynn. Flynn tries to install node and it seems to needs the proxy. So I needed to disable a proxy when starting the cluster, but need a proxy to push an app. Is there a way to do both?

Here's the output:

```text
$ flynn cluster add -p wOixCkDrIxotWkcvy0FD+w/PI1aWa1qbpHCZtUn0n9s= default demo.localflynn.com 7b2c7bd3b3302d580947c9387364d718
Cluster "default" added and set as default.
$ flynn create example
Created example
$ git remote -v
flynn   https://git.demo.localflynn.com/example.git (fetch)
flynn   https://git.demo.localflynn.com/example.git (push)
origin  https://github.com/flynn/nodejs-flynn-example.git (fetch)
origin  https://github.com/flynn/nodejs-flynn-example.git (push)
$ git push flynn master
Counting objects: 25, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (24/24), done.
Writing objects: 100% (25/25), 2.58 KiB | 0 bytes/s, done.
Total 25 (delta 10), reused 0 (delta 0)
-----> Building example...
-----> Node.js app detected

-----> Creating runtime environment

       NPM_CONFIG_LOGLEVEL=error
       NPM_CONFIG_PRODUCTION=true
       NODE_ENV=production
       NODE_MODULES_CACHE=true

-----> Installing binaries
       engines.node (package.json):  0.10.x
       engines.npm (package.json):   2.x

       Resolving node version 0.10.x via semver.io...
       Downloading and installing node ...
       Unable to download node ; does it exist?

-----> Build failed

       We're sorry this build is failing! You can troubleshoot common issues here:
       https://devcenter.heroku.com/articles/troubleshooting-node-deploys

       If you're stuck, please submit a ticket so we can help:
       https://help.heroku.com/

       Love,
       Heroku

remote: Build failed: exec: job exited with status 1
To https://git.demo.localflynn.com/example.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'https://git.demo.localflynn.com/example.git'
```

**Comment by titanous** [comment]
We can configure gitreceive for the proxy separately, but I believe the settings will interfere with some of its communication, so a small patch will be required. Please open a new issue and we'll sort it out.

**Comment by Martin-H1** [comment]
OK I created 1973.
